File: misc/BM2014/BM14_R/EM_DFM_SS_OPT/remNaNs_spline.R

remNaN1o <- function(X, k) {
  d <- dim(X)
  T <- d[1L]
  N <- d[2L]
  k2 <- 2L * k + 1L
  indNaN <- is.na(X)
  wna <- which(indNaN)
  X[wna] <- fmedian(X, TRA = 1L)[wna]
  X_MA <- roll_mean(rbind(X[rep(1L, k), ], X, X[rep(T, k), ]), k2)
  X[wna] <- X_MA[-(1:(k2-1L)), ][wna]
  return(list(X = X, indNaN = indNaN))
}


remNaN2 <- function(X, k, threshold = 0.8) {
  d <- dim(X)
  T <- d[1L]
  N <- d[2L]
  k2 = 2L * k + 1L
  indNaN = is.na(X)
  rem1 = rowSums(indNaN) > N * threshold
  nanLead = cumsum(rem1) == 1:T
  nanEnd = rev(cumsum(rev(rem1)) == 1:T);
  nanLE = nanLead | nanEnd
  X = X[!nanLE, ]
  indNaN = is.na(X)

  for (i in 1:N) {
    x = X[, i]
    nnai = which(!is.na(x))
    ln = length(nnai)
    t1 = nnai[1L]
    t2 = nnai[ln]
    # Cubic spline to interpolate any internal missing values...
    if(ln != t2-t1+1L) x[t1:t2] = spline(nnai, x[nnai], xout = t1:t2)$y
    isnanx = which(is.na(x))
    x[isnanx] = fmedian.default(x)
    x_MA = filter(c(rep(x[1L], k), x, rep(x[T], k)), rep(1, k2)/k2, sides = 1L)
    x[isnanx] = x_MA[-(1:(k2-1L))][isnanx]
    X[, i] = x
  }
  return(list(X = X, indNaN = indNaN))
}

# only remove rows with leading and closing zeros
remNaN3 <- function(X) {
  d <- dim(X)
  T <- d[1L]
  N <- d[2L]
  indNaN = is.na(X)
  rem1 = rowSums(indNaN) == N
  nanLead = cumsum(rem1) == 1:T
  nanEnd = rev(cumsum(rev(rem1)) == 1:T)
  nanLE = nanLead | nanEnd
  X = X[!nanLE, ]
  return(list(X = X, indNaN = is.na(X)))
}

remNaNs <- function(X, options) { # output: [X, indNaN]
  switch(options$method,
         # case 1 % replace all the missing values
         remNaN1o(X, options$k),
         # case 2 % replace missing values after removing leading and closing zeros (= mostly missing rows)
         remNaN2(X, options$k),
         # case 3 % only remove rows with leading and closing zeros
         remNaN3(X),
         stop("remNaN option not implemented yet"))
}

File: misc/BM2014/ExampleSection3/remNaNs_spline.m
function [X,indNaN] = remNaNs(X,options);

% procedure to replace missing observations
% used for ontaining the initial parameters for EM iterations for 
% “MAXIMUM LIKELIHOOD ESTIMATION OF FACTOR MODELS ON DATA SETS WITH ARBITRARY PATTERN OF MISSING DATA.”
% by Marta Ba?bura and Michele Modugno 

[T,N]=size(X);
k=options.k;
indNaN=isnan(X);
switch options.method
    case 1  % replace all the missing values by moving averages (with median)
        for i = 1:N  
            x = X(:,i);
            x(indNaN(:,i))=nanmedian(x);
            x_MA =filter (ones(2*k+1,1)/(2*k+1),1,[x(1)*ones(k,1);x;x(end)*ones(k,1)]);
            x_MA=x_MA(2*k+1:end);
            x(indNaN(:,i))=x_MA(indNaN(:,i));
            X(:,i)=x;
        end
    case 2 %replace missing values after removing leading and closing nans 
        rem1=(sum(indNaN,2)==N);
        nanLead=(cumsum(rem1)==(1:T)');
        nanEnd=(cumsum(rem1(end:-1:1))==(1:T)');
        nanEnd=nanEnd(end:-1:1);
        nanLE=(nanLead | nanEnd);
        X(nanLE,:)=[];
        indNaN=isnan(X);
        for i = 1:N  
            x = X(:,i);
            x(indNaN(:,i))=nanmedian(x);
            x_MA =filter (ones(2*k+1,1)/(2*k+1),1,[x(1)*ones(k,1);x;x(end)*ones(k,1)]);
            x_MA=x_MA(2*k+1:end);
            x(indNaN(:,i))=x_MA(indNaN(:,i));
            X(:,i)=x;
        end
    case 3 %only remove rows with leading and closing nans
        rem1=(sum(indNaN,2)==N);
        nanLead=(cumsum(rem1)==(1:T)');
        nanEnd=(cumsum(rem1(end:-1:1))==(1:T)');
        nanEnd=nanEnd(end:-1:1);
        nanLE=(nanLead | nanEnd);
        X(nanLE,:)=[];
        indNaN = isnan(X);
    case 4 %replace missing values with splines after removing leading and closing nans 
        rem1=(sum(indNaN,2)==N);
        nanLead=(cumsum(rem1)==(1:T)');
        nanEnd=(cumsum(rem1(end:-1:1))==(1:T)');
        nanEnd=nanEnd(end:-1:1);
        nanLE=(nanLead | nanEnd);
        X(nanLE,:)=[];
        indNaN=isnan(X);
        for i = 1:N  
            x = X(:,i);
            isnanx = isnan(x);
            t1 = min(find(~isnanx));
            t2 = max(find(~isnanx));
            x(t1:t2) = spline(find(~isnanx),x(~isnanx),(t1:t2)');
            isnanx = isnan(x);
            x(isnanx)=nanmedian(x);
            x_MA =filter (ones(2*k+1,1)/(2*k+1),1,[x(1)*ones(k,1);x;x(end)*ones(k,1)]);
            x_MA=x_MA(2*k+1:end);
            x(isnanx)=x_MA(isnanx);
            X(:,i)=x;
        end
    case 5 %replace missing values with splines
        indNaN=isnan(X);
        for i = 1:N  
            x = X(:,i);
            isnanx = isnan(x);
            t1 = min(find(~isnanx));
            t2 = max(find(~isnanx));
            x(t1:t2) = spline(find(~isnanx),x(~isnanx),(t1:t2)');
            isnanx = isnan(x);
            x(isnanx)=nanmedian(x);
            x_MA =filter (ones(2*k+1,1)/(2*k+1),1,[x(1)*ones(k,1);x;x(end)*ones(k,1)]);
            x_MA=x_MA(2*k+1:end);
            x(isnanx)=x_MA(isnanx);
            X(:,i)=x;
        end
    case 6 %replace missing values with random draws after removing leading and closing nans 
        rem1=(sum(indNaN,2)==N);
        nanLead=(cumsum(rem1)==(1:T)');
        nanEnd=(cumsum(rem1(end:-1:1))==(1:T)');
        nanEnd=nanEnd(end:-1:1);
        nanLE=(nanLead | nanEnd);
        X(nanLE,:)=[];
        indNaN=isnan(X);
        for i = 1:N  
            x = X(:,i);
            isnanx = isnan(x);
            x(isnanx) = randn(sum(isnanx),1);
            X(:,i)=x;
        end
    case 7 %replace missing values with random draws
        indNaN=isnan(X);
        for i = 1:N  
            x = X(:,i);
            isnanx = isnan(x);
            x(isnanx) = randn(sum(isnanx),1);
            X(:,i)=x;
        end
end
